### DISCUSSION TITLE
[mobx-react] Allow state change when rendering

### DISCUSSION THREAD

**Discussion by namnm** [mobxjs/mobx#3010] [Ideas]
I would like to have an option to allow state change when rendering.

I have an react-native app, and sometimes it crash because of this error. I would like to have an option to disable it to not throw the error.
I can not control the state change now because it is related to api call and timeout/interval.

**Comment by iChenLei** [comment]
Any error log ?

**Reply by namnm** [reply]
I couldnt have the log. We made a fork turning it off manually by:
`rendering = _allowStateChanges(true, baseRender)`
https://github.com/mobxjs/mobx/blob/72d06f8cd2519ce4dbfb807bc13556ca35866690/packages/mobx-react/src/observerClass.ts#L134

Then it fulfill our need.

**Reply by mweststrate** [reply]
> I can not control the state change now because it is related to api call and timeout/interval.

That sounds not like something that is happening during rendering at all, but a bug in how you manage intervals / api calls, or what you do in the handling thereof. If you're experience this error you should debug and fix it. Or build an error boundary to suppress it at your own risk.

**Reply by namnm** [reply]
@mweststrate I guess this happened  because we used `react-native-background-timer` which run the timeout/interval in another separated thread, which can cause the issue updating store during rendering phase.
Moreover it also caused some issue with concurrent map access in some of our native code we fixed those, but for this mobx, we need to make the fork.

**Comment by mweststrate** [comment]
The JS world is still single threaded in RN, so I think your problem lies
elsewhere.

**Reply by namnm** [reply]
Thanks for pointing out. We'll try to investigate more and see where is the root cause. In the meantime we will use the fork to suppress the crash.

Thanks for your support.

**Comment by Epskampie** [comment]
Why don't you just:
```
setTimeout(() => {
 // Do your stuff here
})
```
This way your code is always only run after a tiny delay, when not rendering anymore.
